doc/specs/Authentication.md

```diff
@@ -0,0 +1,473 @@
+* Resetting a password must not disable second-factor authentication.
+* The age of a password should be recorded in the database.
+
+## Use cases
+
+These use cases describe various ways the authentication system will interact.
+
+1. Users need to be able to authenticate by supplying their user name and
+   password. This is the default mode of operation.
+2. The user shows they have the secondary means of authentication.
+3. User allows an application to use their account.
+4. User modifies their password
+5. User configures additional authentication
+6. User reviews authorized applications
+7. User resets their password.
+8. A moderator reviews security settings
+
+### 1. User authenticates with a password
+
+**Primary actor:** User
+
+**Brief:** User supplies their username and password. These are verified and if
+found to be correct the user is identified and logged in to the website.
+
+**Preconditions:**
+* A user has an account on the website with a corresponding password.
+* The user is not currently logged in.
+* The user has not failed to authenticate more than 6 times already.
+* The user is not connecting from a blocked IP-block.
+
+**Trigger:** User tries to access a page but they do not currently have
+an active session.
+
+**Postconditions:**
+* Success:
+  * The user is authenticated using their password.
+  * The failed-authentication count is reset
+  * The log-in is logged to the session log.
+* Failure
+  * The failed-authentication count is incremented.
+
+**Basic flow:**
+
+1. User is presented with a form requesting a username and password
+1. User provides their username and password
+1. User is forwarded to the secondary authentication page if applicable
+   or forwarded to the page they initially tried to visit.
+
+**Alternate flow:**
+
+* Authentication fails:
+  * User is told their credentials were incorrect and to try again.
+  * User is also told how many attempts they have remaining.
+  * User is shown the instructions to retrieve their password.
+* User is locked out:
+  * User is shown the time remaining for the ban to expire.
+  * User is shown the instructions to retrieve their password.
+
+**Additional functional requirements:**
+
+* Always increment the number of attempts, even before checking authentication.
+
+### 2. User uses two-factor authentication
+
+**Primary actor:** User
+
+**Brief:** User uses their source for two-factor authentication to further
+authenticate.
+
+**Preconditions:**
+* User has successfully used a primary means of authentication
+* User has beforehand set up secondary authentication.
+
+**Trigger:** User has successfully authenticated and has set up secondary
+authentication.
+
+**Postcondition:**
+* Success
+  * The user's session is activated.
+  * If a one-time code is used, the code is invalidated and the user
+    is forwarded to the second-factor configuration page.
+* Failure
+  * The user may not proceed and will need to retry.
```

Comment:
Fair question. Typically this will be a new token, as TOTP codes (such as Google Authenticator) expire. However if you'd retry fast enough it's possible that the old token is still valid. That would not really be a problem.

I'm rewording it to 'with a fresh token'